ipld/read.go

```diff
@@ -20,188 +18,130 @@ import (
-		sample := uint32(rand.Intn(max))
-		if _, has := samples[sample]; has {
-			continue
-		}
-		samples[sample] = struct{}{}
-		i++
-	}
-	out := make([]uint32, count)
-	counter := 0
-	for s := range samples {
-		out[counter] = s
-		counter++
-	}
-	return out
-}
-
-type index struct {
-	row uint32
-	col uint32
-}
-
-type indexedShare struct {
-	data []byte
-	index
-}
-
-// shareCounter is a thread safe tallying mechanism for share retrieval
-type shareCounter struct {
-	// all shares
-	shares map[index][]byte
-	// number of shares successfully collected
-	counter uint32
-	// the width of the extended data square
-	edsWidth uint32
-	// the minimum shares needed to repair the extended data square
-	minSharesNeeded uint32
-
-	shareChan chan indexedShare
-	ctx       context.Context
-	cancel    context.CancelFunc
-	// any errors encountered when attempting to retrieve shares
-	errc chan error
+	tree := wrapper.NewErasuredNamespacedMerkleTree(uint64(edsWidth) / 2)
+	return rsmt2d.RepairExtendedDataSquare(rowRoots, colRoots, shares.data, codec, tree.Constructor)
 }
 
-func newshareCounter(parentCtx context.Context, edsWidth uint32) *shareCounter {
-	ctx, cancel := context.WithCancel(parentCtx)
-
-	// calculate the min number of shares needed to repair the square
-	minSharesNeeded := edsWidth * edsWidth / 4
+// fillBlockWithData fetches 1/4 of shares for the given root
+func fillBlockWithData(
+	ctx context.Context,
+	data [][]byte, dag ipld.NodeGetter,
+	isRow bool,
+	f *totalShares,
+) {
+	for i := 0; i < len(data)/2; i++ {
+		go func(i int) {
+			rootHash, err := plugin.CidFromNamespacedSha256(data[i])
+			if err != nil {
+				f.err <- err
+				return
+			}
+			subtreeRootHash, err := getHalfTreeNodeHash(ctx, rootHash, dag, true)
+			if err != nil {
+				f.err <- err
+				return
+			}
 
-	return &shareCounter{
-		shares:          make(map[index][]byte),
-		edsWidth:        edsWidth,
-		minSharesNeeded: minSharesNeeded,
-		shareChan:       make(chan indexedShare, 512),
-		errc:            make(chan error, 1),
-		ctx:             ctx,
-		cancel:          cancel,
+			leafs, err := GetLeafs(ctx, dag, *subtreeRootHash)
+			if err != nil {
+				f.err <- err
+				return
+			}
+			for leafIdx, leaf := range leafs {
+				shareData := leaf.RawData()[1:]
+				// it's not needed to store data for cols
+				// as we are fetching data from the same share for rows and cols
+				if isRow {
+					f.data[(i*len(data))+leafIdx] = shareData[NamespaceSize:]
+				}
+			}
+			f.err <- nil
+		}(i)
 	}
 }
 
-// retrieveLeaf uses GetLeafData to fetch a single leaf and counts that leaf
-func (sc *shareCounter) retrieveShare(
-	rootCid cid.Cid,
-	isRow bool,
-	axisIdx uint32,
-	idx uint32,
-	dag ipld.NodeGetter,
-) {
-	data, err := GetLeafData(sc.ctx, rootCid, idx, sc.edsWidth, dag)
+// getHalfTreeNodeHash returns only one subtree - left or right
+func getHalfTreeNodeHash(ctx context.Context, root cid.Cid, dag ipld.NodeGetter, isLeftSubtree bool) (*cid.Cid, error) {
```

Comment:
You're always calling this with `isLeftSubtree` as true, so why not just have this method return the left subtree only instead of passing in the bool?

Reply:
Can we simplify it for this PR at least to only have this function do `getLeftHalfTreeNodeHash` ? It's cleaner if this is the only usage *for now* IMO.